In [1]:
import tensorflow as tf

from tensorflow.keras.datasets.mnist import load_data

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense

import matplotlib.pyplot as plt

In [2]:
tf.config.list_physical_devices('GPU')

[]

In [3]:
data = load_data()

In [4]:
(x_train, y_train), (x_test, y_test) = data

In [11]:
x_train[0].shape

(28, 28)

In [12]:
y_train[0]

5

In [14]:
x_train = x_train.reshape(len(x_train), -1)
x_test = x_test.reshape(len(x_test), -1)

In [15]:
x_train.shape

(60000, 784)

In [16]:
len(set(y_test))

10

In [21]:
model = Sequential([
    Input(shape = (x_train.shape[1],)),
    Dense(units = 20, activation = "relu"),
    Dense(units = 50, activation = "relu"),
    Dense(units = 30, activation = "relu"),
    Dense(units = len(set(y_train)), activation = "sigmoid")
])

In [22]:
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_4 (Dense)                      │ (None, 20)                  │          15,700 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_5 (Dense)                      │ (None, 50)                  │           1,050 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_6 (Dense)                      │ (None, 30)                  │           1,530 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_7 (Dense)                      │ (None, 10)                  │             310 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 18,590 (72.62 KB)

 Trainable params: 18,590 (72.62 KB)

 Non-trainable params: 0 (0.00 B)

In [25]:
model.compile(loss = "sparse_categorical_crossentropy", optimizer = "adam")

In [24]:
y_train

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [26]:
model.fit(x_train, y_train)

1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 868us/step - loss: 2.1066


In [28]:
model.evaluate(x_train, y_train)

1875/1875 ━━━━━━━━━━━━━━━━━━━━ 1s 612us/step - loss: 0.4203


0.41322818398475647

0.41322818398475647 - това е loss функцията. По-голяма loss функция означава по-голяма грешка.

In [29]:
model.fit(x_train, y_train, epochs = 20)

Epoch 1/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 846us/step - loss: 0.3893
Epoch 2/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 845us/step - loss: 0.3117
Epoch 3/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 841us/step - loss: 0.2826
Epoch 4/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 845us/step - loss: 0.2589
Epoch 5/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 872us/step - loss: 0.2492
Epoch 6/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 869us/step - loss: 0.2283
Epoch 7/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 864us/step - loss: 0.2164
Epoch 8/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 850us/step - loss: 0.2002
Epoch 9/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 861us/step - loss: 0.1992
Epoch 10/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 838us/step - loss: 0.1800
Epoch 11/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 859us/step - loss: 0.1803
Epoch 12/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 842us/step - loss: 0.1704
Epoch 13/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 870us/step - loss: 0.1591
Epoch 14/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 875us/step - loss: 0.1601
E

In [30]:
model.evaluate(x_train, y_train)

1875/1875 ━━━━━━━━━━━━━━━━━━━━ 1s 589us/step - loss: 0.1229


0.12508952617645264